File: BinaryTree_Problems/BTPostOrderTraversal.js

// Problem link - https://leetcode.com/problems/binary-tree-postorder-traversal/

class TreeNode{
    constructor(val){
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

class BinaryTree{
    constructor(){
        this.root = null;
    }

    createBinaryTree(arr, index){
        if(index < arr.length){
            let temp = new TreeNode(arr[index]);
            temp.left = this.createBinaryTree(arr, 2 * index + 1);
            temp.right = this.createBinaryTree(arr, 2 * index + 2);
            return temp;
        }
        return null;
    }

    intializeTreeNode(arr){
        this.root = this.createBinaryTree(arr,0);
    }

    postOrderRec(node, res){
        // base case:
        if (node === null) return;
        // recursive case:
        this.postOrderRec(node.left, res);
        this.postOrderRec(node.right, res);
        res.push(node.val);
    }

    postorderTraversal = function(root) {
        let res = [];
        this.postOrderRec(root, res);
        return res;
    };
}

const bt = new BinaryTree();
let arr = [1,2,3,4,5,6,7];
bt.intializeTreeNode(arr,0);

console.log("Reading binary tree data using postOrder traversal:");
console.log(bt.postorderTraversal(bt.root));

File: BinaryTree_Problems/BTPostOrderTraversal.java

// Problem link - https://leetcode.com/problems/binary-tree-postorder-traversal/

package BinaryTree_Problems;

import java.util.*;

public class BTPostOrderTraversal {
    public static void postOrderRec(TreeNodeUtil node, List<Integer> res){
        // base case:
        if(node == null) return;

        // Recursive case:
        postOrderRec(node.left, res);
        postOrderRec(node.right, res);
        res.add(node.val);
    }
    public static List<Integer> postorderTraversal(TreeNodeUtil root) {
        List<Integer> res = new ArrayList<>();
        postOrderRec(root, res);
        return res;
    }

    public static void main(String[] args) {
        BinaryTreeUtil bt = new BinaryTreeUtil();
        int[] arr = {1, 2, 3, 4, 5, 6, 7};
        bt.initializeTree(arr);

        System.out.println("Reading binary tree data using post Order traversal:");
        List<Integer> res = postorderTraversal(bt.root);
        for(int val : res){
            System.out.print(val + " ");
        }
        System.out.println();
    }
}